src/tree/secret_tree.rs

```diff
@@ -182,38 +213,63 @@ impl ASTree {
             self.hash_down(ciphersuite, n);
         }
         let node_secret = &self.nodes[index_in_tree.as_usize()].clone().unwrap().secret;
-        let mut sender_ratchet = SenderRatchet::new(index, node_secret);
-        let application_secret = sender_ratchet.get_secret(generation, ciphersuite);
+        let handshake_ratchet_secret = derive_tree_secret(
+            ciphersuite,
+            node_secret,
+            "handshake",
+            index.as_u32(),
+            generation,
+            ciphersuite.hash_length(),
+        );
+        let application_ratchet_secret = derive_tree_secret(
+            ciphersuite,
+            node_secret,
+            "application",
+            index.as_u32(),
+            generation,
+            ciphersuite.hash_length(),
+        );
+        let mut handshake_sender_ratchet = SenderRatchet::new(index, &handshake_ratchet_secret);
+        let mut application_sender_ratchet = SenderRatchet::new(index, &application_ratchet_secret);
+        let ratchet_secrets = match secret_type {
+            SecretType::HandshakeSecret => {
+                handshake_sender_ratchet.get_secret(generation, ciphersuite)
+            }
+            SecretType::ApplicationSecret => {
+                application_sender_ratchet.get_secret(generation, ciphersuite)
+            }
+        };
         self.nodes[index_in_tree.as_usize()] = None;
-        self.sender_ratchets[index.as_usize()] = Some(sender_ratchet);
-        application_secret
+        self.handshake_sender_ratchets[index.as_usize()] = Some(handshake_sender_ratchet);
+        self.application_sender_ratchets[index.as_usize()] = Some(application_sender_ratchet);
+        ratchet_secrets
     }
 
     fn hash_down(&mut self, ciphersuite: &Ciphersuite, index_in_tree: NodeIndex) {
         let hash_len = ciphersuite.hash_length();
         let node_secret = &self.nodes[index_in_tree.as_usize()].clone().unwrap().secret;
         let left_index = left(index_in_tree);
         let right_index = right(index_in_tree, self.size);
-        let left_secret = derive_app_secret(
+        let left_secret = derive_tree_secret(
             &ciphersuite,
             &node_secret,
             "tree",
             left_index.as_u32(),
             0,
             hash_len,
```

Comment:
This should be the `KDF.Nh`, but that's a bug in the spec as well. (Same a few lines below.)

Reply:
Will be fixed with #57 and #46.